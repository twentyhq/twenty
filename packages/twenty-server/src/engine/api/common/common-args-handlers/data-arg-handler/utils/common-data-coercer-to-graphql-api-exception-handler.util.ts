import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';
import { UserInputError } from 'src/engine/core-modules/graphql/utils/graphql-errors.util';

export const commonDataCoercerToGraphqlApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new UserInputError(error);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
