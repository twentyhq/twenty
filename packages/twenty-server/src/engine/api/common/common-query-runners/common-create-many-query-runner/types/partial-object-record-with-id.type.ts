import { type ObjectRecord } from 'twenty-shared/types';

export type PartialObjectRecordWithId = Partial<ObjectRecord> & { id: string };
