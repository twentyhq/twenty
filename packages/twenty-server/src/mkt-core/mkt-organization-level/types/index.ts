export * from './organization-level.types';
export * from './permissions.type';
