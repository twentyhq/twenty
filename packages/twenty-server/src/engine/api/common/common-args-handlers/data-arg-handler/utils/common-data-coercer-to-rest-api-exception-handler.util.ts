import { BadRequestException } from '@nestjs/common';

import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';

export const commonDataCoercerToRestApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new BadRequestException(error.message);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
