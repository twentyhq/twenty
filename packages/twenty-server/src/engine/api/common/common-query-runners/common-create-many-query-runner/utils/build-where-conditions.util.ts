import { isDefined } from 'twenty-shared/utils';
import { type FindOperator, In } from 'typeorm';

import { type ObjectRecord } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';

export const buildWhereConditions = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
): Record<string, FindOperator<string>>[] => {
  const whereConditions = [];

  for (const field of conflictingFields) {
    const fieldValues = records
      .map((record) => getValueFromPath(record, field.fullPath))
      .filter(isDefined);

    if (fieldValues.length > 0) {
      whereConditions.push({ [field.column]: In(fieldValues) });
    }
  }

  return whereConditions;
};
