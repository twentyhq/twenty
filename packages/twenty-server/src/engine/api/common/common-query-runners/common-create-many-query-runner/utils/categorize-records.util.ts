import { isDefined } from 'twenty-shared/utils';

import { type ObjectRecord } from 'twenty-shared/types';

import { type PartialObjectRecordWithId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/types/partial-object-record-with-id.type';
import { getMatchingRecordId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-matching-record-id.util';

export const categorizeRecords = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
  existingRecords: PartialObjectRecordWithId[],
): {
  recordsToUpdate: PartialObjectRecordWithId[];
  recordsToInsert: Partial<ObjectRecord>[];
} => {
  const recordsToUpdate: PartialObjectRecordWithId[] = [];
  const recordsToInsert: Partial<ObjectRecord>[] = [];

  for (const record of records) {
    const matchingRecordId = getMatchingRecordId(
      record,
      conflictingFields,
      existingRecords,
    );

    if (isDefined(matchingRecordId)) {
      recordsToUpdate.push({ ...record, id: matchingRecordId });
    } else {
      recordsToInsert.push(record);
    }
  }

  return { recordsToUpdate, recordsToInsert };
};
