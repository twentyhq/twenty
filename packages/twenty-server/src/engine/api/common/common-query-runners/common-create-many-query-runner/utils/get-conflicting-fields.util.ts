import { capitalize } from 'twenty-shared/utils';

import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
import { type ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';

export const getConflictingFields = (
  objectMetadataItemWithFieldMaps: ObjectMetadataItemWithFieldMaps,
): {
  baseField: string;
  fullPath: string;
  column: string;
}[] => {
  return Object.values(objectMetadataItemWithFieldMaps.fieldsById)
    .filter((field) => field.isUnique || field.name === 'id')
    .flatMap((field) => {
      const compositeType = compositeTypeDefinitions.get(field.type);

      if (!compositeType) {
        return [
          {
            baseField: field.name,
            fullPath: field.name,
            column: field.name,
          },
        ];
      }

      const property = compositeType.properties.find(
        (prop) => prop.isIncludedInUniqueConstraint,
      );

      return property
        ? [
            {
              baseField: field.name,
              fullPath: `${field.name}.${property.name}`,
              column: `${field.name}${capitalize(property.name)}`,
            },
          ]
        : [];
    });
};
