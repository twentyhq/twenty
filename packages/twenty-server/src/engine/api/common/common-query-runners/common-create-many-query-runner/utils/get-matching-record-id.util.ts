import { msg } from '@lingui/core/macro';
import { type ObjectRecord } from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';

import { type PartialObjectRecordWithId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/types/partial-object-record-with-id.type';
import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';
import {
  CommonQueryRunnerException,
  CommonQueryRunnerExceptionCode,
} from 'src/engine/api/common/common-query-runners/errors/common-query-runner.exception';

export const getMatchingRecordId = (
  record: Partial<ObjectRecord>,
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
  existingRecords: PartialObjectRecordWithId[],
): string | undefined => {
  const matchingRecordIds = conflictingFields.reduce<string[]>((acc, field) => {
    const requestFieldValue = getValueFromPath(record, field.fullPath);

    const matchingRecord = existingRecords.find((existingRecord) => {
      const existingFieldValue = getValueFromPath(
        existingRecord,
        field.fullPath,
      );

      return (
        isDefined(existingFieldValue) &&
        existingFieldValue === requestFieldValue
      );
    });

    if (isDefined(matchingRecord)) {
      acc.push(matchingRecord.id);
    }

    return acc;
  }, []);

  if ([...new Set(matchingRecordIds)].length > 1) {
    const conflictingFieldsValues = conflictingFields
      .map((field) => {
        const value = getValueFromPath(record, field.fullPath);

        return isDefined(value) ? `${field.fullPath}: ${value}` : undefined;
      })
      .filter(isDefined)
      .join(', ');

    throw new CommonQueryRunnerException(
      `Multiple records found with the same unique field values for ${conflictingFieldsValues}. Cannot determine which record to update.`,
      CommonQueryRunnerExceptionCode.UPSERT_MULTIPLE_MATCHING_RECORDS_CONFLICT,
      {
        userFriendlyMessage: msg`Multiple records found with the same unique field values for ${conflictingFieldsValues}. Cannot determine which record to update.`,
      },
    );
  }

  return matchingRecordIds[0];
};
