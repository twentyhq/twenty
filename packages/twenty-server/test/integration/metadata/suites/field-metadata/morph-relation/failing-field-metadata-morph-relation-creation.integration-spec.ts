import { faker } from '@faker-js/faker/.';
import { createOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util';
import { createOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/create-one-object-metadata.util';
import { deleteOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { getMockCreateObjectInput } from 'test/integration/metadata/suites/object-metadata/utils/generate-mock-create-object-metadata-input';
import { EachTestingContext } from 'twenty-shared/testing';
import { FieldMetadataType } from 'twenty-shared/types';

import { RelationType } from 'src/engine/metadata-modules/field-metadata/interfaces/relation-type.interface';

import { CreateFieldInput } from 'src/engine/metadata-modules/field-metadata/dtos/create-field.input';

type GlobalTestContext = {
  objectMetadataIds: {
    firstTargetObjectId: string;
    secondTargetObjectId: string;
    sourceObjectId: string;
  };
  targetFieldLabel: string;
  type: FieldMetadataType;
  targetFieldIcon: string;
  collisionFieldLabel: string;
};
const globalTestContext: GlobalTestContext = {
  objectMetadataIds: {
    firstTargetObjectId: '',
    secondTargetObjectId: '',
    sourceObjectId: '',
  },
  targetFieldLabel: 'defaultTargetFieldLabel',
  type: FieldMetadataType.MORPH_RELATION,
  targetFieldIcon: 'IconBuildingSkyscraper',
  collisionFieldLabel: 'collisionfieldlabel',
};

type TestedRelationCreationPayload = Partial<
  NonNullable<CreateFieldInput['relationCreationPayload']>
>;

type CreateOneObjectMetadataItemTestingContext = EachTestingContext<
  | TestedRelationCreationPayload
  | ((context: GlobalTestContext) => TestedRelationCreationPayload)
>[];
describe('Field metadata morph relation creation should fail', () => {
  const failingLabelsCreationTestsUseCase: CreateOneObjectMetadataItemTestingContext =
    [
      {
        title: 'when targetFieldLabel is empty',
        context: { targetFieldLabel: '' },
      },
      {
        title: 'when targetFieldLabel exceeds maximum length',
        context: {
          targetFieldLabel: 'A'.repeat(64),
        },
      },
      {
        // Not handled gracefully should be refactored
        title: 'when targetObjectMetadataId is unknown',
        context: {
          targetObjectMetadataId: faker.string.uuid(),
        },
      },
      {
        title: 'when targetFieldLabel contains only whitespace',
        context: { targetFieldLabel: '   ' },
      },
      {
        title:
          'when targetFieldLabel conflicts with an existing field on target object metadata id',
        context: ({ collisionFieldLabel }) => ({
          targetFieldLabel: collisionFieldLabel,
        }),
      },
      {
        title: 'when type is not provided',
        context: { type: undefined },
      },
      {
        title: 'when type is a wrong value',
        context: { type: 'wrong' as RelationType },
      },
    ];

  beforeAll(async () => {
    const {
      data: {
        createOneObject: { id: sourceObjectId },
      },
    } = await createOneObjectMetadata({
      input: getMockCreateObjectInput({
        namePlural: 'sourceRelations',
        nameSingular: 'sourceRelation',
      }),
    });

    const {
      data: {
        createOneObject: { id: firstTargetObjectId },
      },
    } = await createOneObjectMetadata({
      input: getMockCreateObjectInput({
        namePlural: 'firstTargetRelations',
        nameSingular: 'firstTargetRelation',
      }),
    });

    const {
      data: {
        createOneObject: { id: secondTargetObjectId },
      },
    } = await createOneObjectMetadata({
      input: getMockCreateObjectInput({
        namePlural: 'secondTargetRelations',
        nameSingular: 'secondTargetRelation',
      }),
    });

    globalTestContext.objectMetadataIds = {
      sourceObjectId,
      firstTargetObjectId,
      secondTargetObjectId,
    };

    const { data } = await createOneFieldMetadata({
      input: {
        objectMetadataId: firstTargetObjectId,
        name: globalTestContext.collisionFieldLabel,
        label: 'LabelThatCouldBeAnything',
        isLabelSyncedWithName: false,
        type: FieldMetadataType.TEXT,
      },
    });

    expect(data).toBeDefined();
  });

  afterAll(async () => {
    for (const objectMetadataId of Object.values(
      globalTestContext.objectMetadataIds,
    )) {
      await deleteOneObjectMetadata({
        input: {
          idToDelete: objectMetadataId,
        },
      });
    }
  });

  it.each(failingLabelsCreationTestsUseCase)(
    'relation ONE_TO_MANY $title',
    async ({ context }) => {
      const computedContext =
        typeof context === 'function' ? context(globalTestContext) : context;

      const { errors } = await createOneFieldMetadata({
        expectToFail: true,
        input: {
          objectMetadataId: globalTestContext.objectMetadataIds.sourceObjectId,
          name: 'owner',
          label: 'owner field',
          isLabelSyncedWithName: false,
          type: FieldMetadataType.MORPH_RELATION,
          morphRelationsCreationPayload: [
            {
              targetFieldLabel: 'defaultFirstTargetFieldLabel',
              type: RelationType.ONE_TO_MANY,
              targetObjectMetadataId:
                globalTestContext.objectMetadataIds.firstTargetObjectId,
              targetFieldIcon: 'IconBuildingSkyscraper',
              ...computedContext,
            },
            {
              targetFieldLabel: 'defaultSecondTargetFieldLabel',
              type: RelationType.ONE_TO_MANY,
              targetObjectMetadataId:
                globalTestContext.objectMetadataIds.secondTargetObjectId,
              targetFieldIcon: 'IconBuildingSkyscraper',
              ...computedContext,
            },
          ],
        },
      });

      expect(errors).toBeDefined();
      expect(errors).toMatchSnapshot();
    },
  );

  it.each(failingLabelsCreationTestsUseCase)(
    'relation MANY_TO_ONE $title',
    async ({ context }) => {
      const computedContext =
        typeof context === 'function' ? context(globalTestContext) : context;

      const { errors } = await createOneFieldMetadata({
        expectToFail: true,
        input: {
          objectMetadataId: globalTestContext.objectMetadataIds.sourceObjectId,
          name: 'owner',
          label: 'owner field',
          isLabelSyncedWithName: false,
          type: FieldMetadataType.MORPH_RELATION,
          morphRelationsCreationPayload: [
            {
              targetFieldLabel: 'defaultFirstTargetFieldLabel',
              type: RelationType.MANY_TO_ONE,
              targetObjectMetadataId:
                globalTestContext.objectMetadataIds.firstTargetObjectId,
              targetFieldIcon: 'IconBuildingSkyscraper',
              ...computedContext,
            },
            {
              targetFieldLabel: 'defaultSecondTargetFieldLabel',
              type: RelationType.MANY_TO_ONE,
              targetObjectMetadataId:
                globalTestContext.objectMetadataIds.secondTargetObjectId,
              targetFieldIcon: 'IconBuildingSkyscraper',
              ...computedContext,
            },
          ],
        },
      });

      expect(errors).toBeDefined();
      expect(errors).toMatchSnapshot();
    },
  );
});
