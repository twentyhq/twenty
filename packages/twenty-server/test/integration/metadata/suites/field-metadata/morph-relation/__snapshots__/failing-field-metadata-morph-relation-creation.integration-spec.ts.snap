// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field metadata morph relation creation should fail relation MANY_TO_ONE when targetFieldLabel conflicts with an existing field on target object metadata id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Name "collisionfieldlabel" is not available",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation MANY_TO_ONE when targetFieldLabel contains only whitespace 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Invalid label: "   "",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation MANY_TO_ONE when targetFieldLabel exceeds maximum length 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Name is too long: it exceeds the 63 characters limit.",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation MANY_TO_ONE when targetFieldLabel is empty 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Input is too short: """,
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation MANY_TO_ONE when targetObjectMetadataId is unknown 1`] = `
[
  {
    "extensions": {
      "code": "INTERNAL_SERVER_ERROR",
      "exceptionEventId": "mocked-exception-id",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Object metadata relation target not found for relation creation payload",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation MANY_TO_ONE when type is a wrong value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Relation creation payload is invalid: type must be one of the following values: ONE_TO_MANY, MANY_TO_ONE",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation MANY_TO_ONE when type is not provided 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Relation creation payload is invalid: type must be one of the following values: ONE_TO_MANY, MANY_TO_ONE",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation ONE_TO_MANY when targetFieldLabel conflicts with an existing field on target object metadata id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Name "collisionfieldlabel" is not available",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation ONE_TO_MANY when targetFieldLabel contains only whitespace 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Invalid label: "   "",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation ONE_TO_MANY when targetFieldLabel exceeds maximum length 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Name is too long: it exceeds the 63 characters limit.",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation ONE_TO_MANY when targetFieldLabel is empty 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Input is too short: """,
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation ONE_TO_MANY when targetObjectMetadataId is unknown 1`] = `
[
  {
    "extensions": {
      "code": "INTERNAL_SERVER_ERROR",
      "exceptionEventId": "mocked-exception-id",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Object metadata relation target not found for relation creation payload",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation ONE_TO_MANY when type is a wrong value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Relation creation payload is invalid: type must be one of the following values: ONE_TO_MANY, MANY_TO_ONE",
    "name": "UserInputError",
  },
]
`;

exports[`Field metadata morph relation creation should fail relation ONE_TO_MANY when type is not provided 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Relation creation payload is invalid: type must be one of the following values: ONE_TO_MANY, MANY_TO_ONE",
    "name": "UserInputError",
  },
]
`;
